runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,477 @@
+use crate::{Error, Spawner};
+use commonware_macros::select;
+use commonware_utils::{from_hex, hex};
+use futures::{
+    channel::{mpsc, oneshot},
+    future::Either,
+    SinkExt as _, StreamExt as _,
+};
+use io_uring::{opcode, squeue::Entry as SqueueEntry, types, IoUring};
+use std::{
+    collections::HashMap,
+    fs::{self, File},
+    future::Future,
+    io::{Error as IoError, ErrorKind},
+    os::fd::{AsRawFd as _, OwnedFd},
+    path::PathBuf,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+const IOURING_SIZE: u32 = 128;
+
+#[derive(Clone)]
+pub struct Config {
+    pub storage_directory: PathBuf,
+}
+
+impl Config {
+    pub fn new(storage_directory: PathBuf) -> Self {
+        Self { storage_directory }
+    }
+}
+
+/// A future that resolves to the next completion queue entry when available
+struct NextCompletionFuture<'a> {
+    ring: &'a mut IoUring,
+}
+
+impl<'a> NextCompletionFuture<'a> {
+    fn new(ring: &'a mut IoUring) -> Self {
+        Self { ring }
+    }
+}
+
+impl Future for NextCompletionFuture<'_> {
+    type Output = io_uring::cqueue::Entry;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Try to get a completion
+        if let Some(cqe) = self.ring.completion().next() {
+            return Poll::Ready(cqe);
+        }
+
+        // Submit any pending operations, which might generate completions
+        self.ring.submit().expect("unable to submit to ring");
+
+        // Try again after submitting
+        if let Some(cqe) = self.ring.completion().next() {
+            return Poll::Ready(cqe);
+        }
+
+        // No completions yet, register waker and return Pending
+        cx.waker().wake_by_ref();
```

Comment:
The runtime will automatically attempt to wake the future again when returning `Poll::Pending`?

If we were to do this, we should just put the loop in `spawn_blocking` and never yield.